File: cli/c++/player.cpp
/*
Blackjack 

Copyright 2015 Sam Saint-Pettersen
Released under the MIT/X11 License.
*/

#include <string>
#include <vector>
#include <cstdio>
#include "cards.h"
#include "split.h"
using namespace std;

class Player {
private:
	vector<string> cards;
	vector<int> values;
	int calcTotal();
public:
	string hit(Cards cards);
	void stand();
	bool hasBlackjack();
	bool isBust();
	vector<string> receiveCards(vector<string> player_cards);
	int showCards();
};

int Player::calcTotal() {
	int total = 0;
	for(int i = 0; i < values.size(); i++) {
		int v = values[i];
		if(v == 1) {
			if((total + 11) <= 21) v = 11;
			if((total + 11) > 21) v = 1;
		}
		total += v;
	}
	return total;
}

string Player::hit(Cards cards) {
	string card = cards.draw();
	this->cards.push_back(card);
	this->values.push_back(cards.getValue());
	cout << "Player hits." << endl;
	cout << "Player gets " << card << endl;
	return card;
}

void Player::stand() {
	cout << "Player stands." << endl;
}

bool Player::hasBlackjack() {
	bool blackjack = false;
	if(calcTotal() == 21) {
		blackjack = true;
	}
	return blackjack;
}

bool Player::isBust() {
	bool bust = false;
	if(calcTotal() > 21) {
		bust = true;
	}
	return bust;
}

vector<string> Player::receiveCards(vector<string> player_cards) {
	string cards = "";
	for(int i = 0; i < player_cards.size(); i++) {
		vector<string> cv = split(player_cards[i], ':');
		this->cards.push_back(cv[0]);
		this->values.push_back(atoi(cv[1].c_str()));
	}
	cards += this->cards[0] + this->cards[1];
	cout << "\nPlayer receives their cards:" << endl;
	cout << cards << " --> " << calcTotal() << endl;
	return this->cards;
}

File: cli/python/player.py
"""
Blackjack

Copyright 2015 Sam Saint-Pettersen
Released under the MIT/X11 License.

Player class
"""
from termcolor import cprint

class Player:

	use_color = True
	cards = []
	values = []

	def __init__(self, color):
		Player.use_color = color
		Player.cards = []
		Player.values = []

	def calcTotal(self):
		Player.values.sort(reverse=True)
		total = 0
		for v in Player.values:
			if v == 1:
				if total + 11 <= 21: v = 11
				if total + 11 > 21: v = 1
			total += v
		return total

	def hasBlackjack(self):
		if self.calcTotal() == 21:
			_print('\nPlayer has Blackjack!', 'magenta')
			return True

	def isBust(self):
		if self.calcTotal() > 21:
			_print('\nPlayer is bust!', 'magenta')
			return True
		
	def receiveCards(self, player_cards):
		cards = ''
		for card in player_cards:
			cv = card.split(':')
			Player.cards.append(cv[0])
			Player.values.append(int(cv[1]))
		
		cards += Player.cards[0] + Player.cards[1]
		_print('Player receives their cards:', 'magenta')
		_print('{0} --> {1}'.format(cards, self.calcTotal()), '')

	def hit(self, cards):
		card = cards.draw()
		Player.cards.append(card)
		_print('Player hits.', 'magenta')
		_print('Player gets ' + card, 'magenta')
		Player.values.append(cards.getValue())
		_print('Player has ' + str(self.calcTotal()), 'magenta')

	def stand(self):
		_print('Player stands.', 'magenta')
		_print('Player has ' + str(self.calcTotal()), 'magenta')

	def showCards(self):
		cards = ''
		for card in Player.cards:
			cards += card

		_print('\nPlayer has:', 'magenta')
		_print('{0} --> {1}'.format(cards, self.calcTotal()), '')
		return self.calcTotal()

def _print(message, color):
	if Player.use_color and color != '':
		cprint(message, color)
	else:
		print(message)

